theano/tensor/basic.py

```diff
@@ -6039,6 +6039,54 @@ def outer(x, y):
         y.dimshuffle('x', 0))
 
 
+def _matmul2(x, y):
+    """numpy.matmul() equivalent for 2D-or-more tensors only."""
+
+    # Make number of dimensions match, prepend broadcastable axes if needed.
+    newdims = abs(x.ndim - y.ndim)
+    if x.ndim < y.ndim:
+        x = x.dimshuffle(*(['x'] * newdims + list(range(x.ndim))))
+        dims = y.ndim
+    else:
+        y = y.dimshuffle(*(['x'] * newdims + list(range(y.ndim))))
+        dims = x.ndim
+
+    # Align the last axis of @x with the last second axis of @y
+    x = x.dimshuffle(*(list(range(dims)) + ['x']))
+    y = y.dimshuffle(*(list(range(dims - 2)) + ['x', dims - 2, dims - 1]))
+
+    # Compute the "elementwise" matrix product
+    return (x * y).sum(axis=-2)
```

Comment:
I guess you're right.  I profiled my matmul with numpy's and found that mine is more than 2 times slower:
```
In [9]: %timeit NP.matmul(RNG.randn(16, 20, 100, 50), RNG.randn(16, 20, 50, 100))
1 loop, best of 3: 277 ms per loop

In [10]: %timeit f(RNG.randn(16, 20,100, 50), RNG.randn(16, 20, 50, 100))
1 loop, best of 3: 736 ms per loop
```

I'll take a look at how `BatchedDot` works and see if I can do `matmul` more efficiently using standard BLAS.  I think there should be better ways doing this.